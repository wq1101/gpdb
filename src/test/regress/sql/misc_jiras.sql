drop schema if exists misc_jiras;
create schema misc_jiras;

--
-- Test backward scanning of tuplestore spill files.
--
-- When tuplestore cannot store all the data in memory it will spill some of
-- the data to temporary files.  In gpdb we used to disable the backward
-- scanning from these spill files because we could not determine the tuple
-- type, memtup or heaptup, correctly.  The issue is fixed, the backward
-- scanning should be supported now.
--

create table misc_jiras.t1 (c1 int, c2 text, c3 smallint) distributed by (c1);
insert into misc_jiras.t1 select i % 13, md5(i::text), i % 3
  from generate_series(1, 20000) i;

-- tuplestore uses work_mem to control the in-memory data size, set a small
-- value to trigger the spilling.
set work_mem to '64kB';

select sum(cc) from (
    select c1
         , c2
         , case when count(c3) = 0 then -1.0
                else cume_dist() over (partition by c1,
                                       case when count(c3) > 0 then 1 else 0 end
                                       order by count(c3), c2)
           end as cc
      from misc_jiras.t1
     group by 1, 2
) tt;

reset work_mem;
