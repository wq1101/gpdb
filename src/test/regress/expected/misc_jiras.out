drop schema if exists misc_jiras;
create schema misc_jiras;
--
-- Test backward scanning of tuplestore spill files.
--
-- When tuplestore cannot store all the data in memory it will spill some of
-- the data to temporary files.  In gpdb we used to disable the backward
-- scanning from these spill files because we could not determine the tuple
-- type, memtup or heaptup, correctly.  The issue is fixed, the backward
-- scanning should be supported now.
--
create table misc_jiras.t1 (c1 int, c2 text, c3 smallint) distributed by (c1);
insert into misc_jiras.t1 select i % 13, md5(i::text), i % 3
  from generate_series(1, 20000) i;
-- tuplestore uses work_mem to control the in-memory data size, set a small
-- value to trigger the spilling.
set work_mem to '64kB';
WARNING:  "work_mem": setting is deprecated, and may be removed in a future release.
select sum(cc) from (
    select c1
         , c2
         , case when count(c3) = 0 then -1.0
                else cume_dist() over (partition by c1,
                                       case when count(c3) > 0 then 1 else 0 end
                                       order by count(c3), c2)
           end as cc
      from misc_jiras.t1
     group by 1, 2
) tt;
   sum   
---------
 10006.5
(1 row)

reset work_mem;
WARNING:  "work_mem": setting is deprecated, and may be removed in a future release.
